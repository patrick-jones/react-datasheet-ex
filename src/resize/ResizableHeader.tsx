import React, {PureComponent, ReactNode} from 'react';
import {ResizableHeaderProps} from './interfaces';
import {ColumnHeader, HeaderRendererProps} from '../interfaces';
import HeaderTitle from '../HeaderTitle';
import ResizableHeaderContainer from './ResizableHeaderContainer';


export interface ResizableHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends ResizableHeaderProps, HeaderRendererProps<T> {
}

export default class ResizableHeader
  extends PureComponent<ResizableHeaderRendererProps> {

  static defaultProps = {
    cellTag: 'TH',
    minSize: 17,
  };

  render(): ReactNode {
    const {index, onResizeStart, onResizing, onResizeEnd, cellTag, minSize, ...rest} = this.props;
    return (
      <ResizableHeaderContainer
        index={index}
        onResizeStart={onResizeStart}
        onResizing={onResizing}
        onResizeEnd={onResizeEnd}
        cellTag={cellTag}
        minSize={minSize}
      >
        <HeaderTitle {...rest} />
      </ResizableHeaderContainer>
    );
  }
}
