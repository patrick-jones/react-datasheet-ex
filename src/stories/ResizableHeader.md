# ResizableHeader

This is the simplest but least flexible component. It adds a drag handle to the right
side of every header, but otherwise acts like a `HeaderTitle` component. This component
requires all of the properties from `HeaderRendererProps` and well as `ResizableProps`.

<!-- STORY -->


## Sample Code

To render the default `title` property:

```jsx
import {ClassNames, Row, Sheet} from 'react-datasheet-ex;
import {ResizableHeader} from 'react-datasheet-ex/resize';

class MyClass extends Component {
  
  headerRenderer = (headerProps) => (
    <ResizableHeader
      index={props.col}
      onResizing={this.props.actions.headerResizing}
      onResizeEnd={this.props.actions.headerResizeEnd}
      {...headerProps}
    />
  );
    
  // ...rest of class...  

  render(): ReactNode {

    const {actions, model: {rows, selected}} = this.props;
    const data = this.getData(rows);

    return (
      <ScrollPanel height={120} width={450}>
        <ReactDatasheet
          className={ClassNames.DATA_GRID_RESIZABLE}
          data={data}
          sheetRenderer={this.sheetRenderer}
          rowRenderer={this.rowRenderer}
          valueRenderer={renderValue}
          onCellsChanged={actions.cellsChanged}
          onSelect={actions.selectionChanged}
          selected={selected}
        />
      </ScrollPanel>
    );
  }
}
```

To render something besides `title`:

```jsx
class MyClass extends Component {
  
  headerRenderer = (headerProps) => (
    <ResizableHeader
      index={props.col}
      onResizing={this.props.actions.headerResizing}
      onResizeEnd={this.props.actions.headerResizeEnd}
      {...headerProps}
    >
      {whateverYouWant}
    </ResizableHeader>
  );
    
  // ...rest of class...
}
```
