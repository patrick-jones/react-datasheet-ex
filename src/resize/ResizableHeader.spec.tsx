import React from 'react';
import 'jest';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import ResizableHeader from './ResizableHeader';


describe('ResizableHeader', () => {
  it('Should render', () => {
    const end = jest.fn();
    const wrapper = mount(
      <table>
        <thead>
          <tr>
            <th>
              <ResizableHeader onResizeEnd={end} index={0} col={0} header={{id: 'A', title: 'A'}}>
                Custom
              </ResizableHeader>
            </th>
            <th>
              <ResizableHeader onResizeEnd={end} index={0} col={0} header={{id: 'B', title: 'B'}} />
            </th>
          </tr>
        </thead>
      </table>
    );

    expect(wrapper.find(ResizableHeader).first().text()).toBe('Custom');
    expect(wrapper.find(ResizableHeader).last().text()).toBe('B');

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });
});
