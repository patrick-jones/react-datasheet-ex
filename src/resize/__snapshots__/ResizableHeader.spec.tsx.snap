// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableHeader Should render 1`] = `
<table>
  <thead>
    <tr>
      <th>
        <ResizableHeader
          cellTag="TH"
          col={0}
          header={
            Object {
              "id": "A",
              "title": "A",
            }
          }
          index={0}
          minSize={17}
          onResizeEnd={[MockFunction]}
        >
          <ResizableHeaderContainer
            cellTag="TH"
            index={0}
            minSize={17}
            onResizeEnd={[MockFunction]}
          >
            <HeaderContainer
              className=""
            >
              <span
                className="rdx-header-cell-container"
              >
                <HeaderTitle
                  className=""
                  col={0}
                  header={
                    Object {
                      "id": "A",
                      "title": "A",
                    }
                  }
                >
                  <span
                    className="rdx-header-cell-title rdx-header-cell-title--clip"
                  >
                    Custom
                  </span>
                </HeaderTitle>
                <WidthResizeHandle
                  cellTag="TH"
                  index={0}
                  minSize={17}
                  onResizeEnd={[MockFunction]}
                >
                  <span
                    className="rdx-width-resize-handle"
                    onMouseDown={[Function]}
                  />
                </WidthResizeHandle>
              </span>
            </HeaderContainer>
          </ResizableHeaderContainer>
        </ResizableHeader>
      </th>
      <th>
        <ResizableHeader
          cellTag="TH"
          col={0}
          header={
            Object {
              "id": "B",
              "title": "B",
            }
          }
          index={0}
          minSize={17}
          onResizeEnd={[MockFunction]}
        >
          <ResizableHeaderContainer
            cellTag="TH"
            index={0}
            minSize={17}
            onResizeEnd={[MockFunction]}
          >
            <HeaderContainer
              className=""
            >
              <span
                className="rdx-header-cell-container"
              >
                <HeaderTitle
                  className=""
                  col={0}
                  header={
                    Object {
                      "id": "B",
                      "title": "B",
                    }
                  }
                >
                  <span
                    className="rdx-header-cell-title rdx-header-cell-title--clip"
                  >
                    B
                  </span>
                </HeaderTitle>
                <WidthResizeHandle
                  cellTag="TH"
                  index={0}
                  minSize={17}
                  onResizeEnd={[MockFunction]}
                >
                  <span
                    className="rdx-width-resize-handle"
                    onMouseDown={[Function]}
                  />
                </WidthResizeHandle>
              </span>
            </HeaderContainer>
          </ResizableHeaderContainer>
        </ResizableHeader>
      </th>
    </tr>
  </thead>
</table>
`;
